apps/ios/SimpleX NSE/NotificationService.swift

```diff
@@ -219,51 +266,87 @@ class NotificationService: UNNotificationServiceExtension {
         deliverBestAttemptNtf()
     }
 
+    func addExpectedMessage(ntfMessage: UserNtfMessage) {
+        if let connEntity = ntfMessage.connEntity_,
+           let receiveEntityId = connEntity.id, ntfMessage.expectedMsg_ != nil {
+            let expectedMsgId = ntfMessage.expectedMsg_?.msgId
+            let receivedMsgId = ntfMessage.receivedMsg_?.msgId
+            logger.debug("NotificationService: addExpectedMessage: expectedMsgId = \(expectedMsgId ?? "nil", privacy: .public), receivedMsgId = \(receivedMsgId ?? "nil", privacy: .public)")
+            expectedMessages[receiveEntityId] = ExpectedMessage(
+                ntfMessage: ntfMessage,
+                receiveConnId: connEntity.conn.agentConnId,
+                expectedMsgId: expectedMsgId,
+                allowedGetNextAttempts: 3,
+                msgBestAttempt: (
+                    ntfMessage.ntfsEnabled
+                    ? .nseData(.connectionEvent(ntfMessage.user, connEntity))
+                    : .empty
+                ),
+                ready: false
+            )
+        }
+    }
+
     override func serviceExtensionTimeWillExpire() {
         logger.debug("DEBUGGING: NotificationService.serviceExtensionTimeWillExpire")
         deliverBestAttemptNtf(urgent: true)
     }
 
-    func processReceivedNtf(_ ntf: NSENotification) -> Bool {
-        guard let ntfInfo = notificationInfo, let expectedMsgTs = ntfInfo.expectedMsg_?.msgTs else { return false }
+    // TODO
+    // for cases where we couldn't process notification:
+    // - return `!expectedMessages.allSatisfy { $0.value.ready }` instead of false?
+    //   (so that outside loop goes on to expected messages from other connections)
+    // - or instead of waiting in outside loop, do one more attempt, if another matching thread is not found proceed?
```

Comment:
Currently just returning `!expectedMessages.allSatisfy { $0.value.ready }`, which means message loop continues while there are not ready "expected message" entries, and that messages previously returning false do not get processed by other threads if they exist. Simply returning false is not correct because then if there are no other matching threads loop will be stuck on waiting for a single message, and batch of expected messages would not finish processing.

Couple of options:
- We could change return to some enum instead of bool signalling "go onto next", "limited retry" (e.g. when not all expected messages are ready - try once or a few more times to find another thread, but shorter than current wait limit), "wait" (when all expected messages are ready - ok to wait as before).
- We could put these rejected messages to some other queue where they would be waiting for matching threads as before, without holding processing of this batch of expected messages.